src/net/FetchHttpClient.ts

```diff
@@ -15,19 +21,72 @@ import {
  * node-fetch package (https://github.com/node-fetch/node-fetch).
  */
 export class FetchHttpClient extends HttpClient implements HttpClientInterface {
-  _fetchFn?: typeof fetch;
+  private readonly _fetchFn: FetchWithTimeout;
 
-  constructor(fetchFn?: typeof fetch) {
+  constructor(fetchFn = fetch) {
     super();
-    this._fetchFn = fetchFn;
+    // Both timeout behaviors differs from Node:
+    // - Fetch uses a single timeout for the entire length of the request.
+    // - Node is more fine-grained and resets the timeout after each stage of the request.
+    if (globalThis.AbortController) {
+      // Utilise native AbortSignal if available
+      // AbortController was added in Node v15.0.0, v14.17.0
+      this._fetchFn = FetchHttpClient.makeFetchWithAbortTimeout(fetch);
+    } else {
+      // Fall to racing against a timeout promise if not available in the runtime
+      // This does not actually cancel the underlying fetch operation or resources
+      this._fetchFn = FetchHttpClient.makeFetchWithRaceTimeout(fetchFn);
+    }
+  }
+
+  private static makeFetchWithRaceTimeout(
+    fetchFn: typeof fetch
+  ): FetchWithTimeout {
+    return (url, init, timeout): Promise<Response> => {
+      let pendingTimeoutId: NodeJS.Timeout | null;
+      const timeoutPromise = new Promise<never>((_, reject) => {
+        pendingTimeoutId = setTimeout(() => {
+          pendingTimeoutId = null;
+          reject(HttpClient.makeTimeoutError());
+        }, timeout);
+      });
+
+      const fetchPromise = fetchFn(url, init);
+      return Promise.race([fetchPromise, timeoutPromise]).finally(() => {
+        if (pendingTimeoutId) {
+          clearTimeout(pendingTimeoutId);
+        }
+      });
+    };
+  }
+
+  private static makeFetchWithAbortTimeout(
+    fetchFn: typeof fetch
+  ): FetchWithTimeout {
+    return (url, init, timeout: number): Promise<Response> => {
+      // Use AbortController because AbortSignal.timeout() was added later in Node v17.3.0, v16.14.0
+      const abort = new AbortController();
+      let timeoutId: NodeJS.Timeout | null = setTimeout(() => {
+        timeoutId = null;
+        abort.abort(HttpClient.makeTimeoutError());
+      }, timeout);
+      return fetchFn(url, {
+        ...init,
+        signal: AbortSignal.timeout(timeout),
```

Comment:
Considering your comment above, should we guard against AbortSignal absence?

Reply:
Sorry mate that one was a bit embarassing! It's supposed to be using the `AbortController.signal` I set up, not `AbortSignal.timeout()`. Fixed that.